review

ornella-mansilla opened the pull request.

ornellalujan committed:
prueba

ornellalujan committed:
Merge pull request #1 from ornellalujan/devBranch

prueba

ornellalujan committed:
prueba 2

ornellalujan committed:
Merge pull request #1 from ornellalujan/devBranch

Dev branch

ornellalujan committed:
Merge pull request #2 from ornellalujan/main

queriendo pasar data de mi fork al original

ornella-mansilla committed:
prueba 3

ornella-mansilla committed:
Merge branch 'main' into dev

ornella-mansilla committed:
Merge pull request #3 from ornella-mansilla/dev

prueba 3

ornella-mansilla committed:
prueba final

ornella-mansilla committed:
Merge pull request #4 from ornella-mansilla/dev

prueba final

ornellalujan committed:
patron state + logica de adivinazas y resolución

ornellalujan committed:
Merge pull request #2 from ornellalujan/devBranch

Dev branch

ornellalujan committed:
Merge pull request #5 from ornellalujan/main

cambios state pattern + logica de adivinanzas y resolucion

ornella-mansilla committed:
Version 1 of Minefield, from dev

Version 1 of Minefield

didierrenardgb commented on minefield_game/board_header.h:
```diff
@@ -9,9 +9,7 @@ struct Board
 
 };
-//void showBoard(const Board& board);
 namespace BoardUtils
```
#should preferably use dromedaryCase instead of CamelCase for namespaces. CamelCase is most widely used for classes and structs, so it'd be easier to confuse with one of those.

didierrenardgb commented on minefield_game/board_header.h:
```diff
@@ -9,9 +9,7 @@ struct Board
     std::vector<std::vector<Cell>> grid;
```
Adding to this, it becomes even more confusing that you have a `board_header.h` without a `board_header.cpp`, and a `board.cpp` without a `board.h`.

didierrenardgb commented on minefield_game/board.cpp:
```diff
@@ -2,88 +2,11 @@
+        // print column headers
         std::cout << "    ";
         for (int x : board.axisX)
         {
```
Cannot comment on line 16, but #should try to avoid braceless conditional and iteration constructs. It's fine if that's your style and wish to keep it that way, but I strongly advice to always use braces on all of those things, it's less error prone.

didierrenardgb commented on minefield_game/board.cpp:
```diff
@@ -100,7 +23,7 @@ namespace BoardUtils
-                // Chequeo si esta celda es mina del jugador
+                // Check if this cell is/has a player's mine
                 bool isMineForPlayer = false;
                 for (const auto& mine : player.mines)
```
#could just write `const Mine& mine`; `auto` and `Mine` have the same amount of characters 😅

didierrenardgb commented on minefield_game/cell.cpp:
```diff
@@ -1,46 +1,14 @@
+void Cell::setState(CellStatus newState) { state = newState; }
+CellStatus Cell::getState() const { return state; }
+bool Cell::isTaken() const { return state == CellStatus::Taken; }
 char Cell::getSymbol() const { return static_cast<char>(state); }
```
#nice usage of `static_cast`!

didierrenardgb commented on minefield_game/cell.cpp:
```diff
@@ -1,46 +1,14 @@
 Cell::Cell(int x, int y)
-    : x(x), y(y), state(CellState::Empty){}
-Cell::Cell(int x, int y, CellState state)
+    : x(x), y(y), state(CellStatus::Empty){}
+Cell::Cell(int x, int y, CellStatus state)
     : x(x), y(y), state(state) {}
```
#nice use of the initialization section!

didierrenardgb commented on minefield_game/ejercicio.exe:
#should not upload generated binaries to the repo

But fine to keep it if it's due to the same reason as Franco!

didierrenardgb commented on minefield_game/init.cpp:
```diff
@@ -35,69 +35,89 @@ namespace Init
+            if (!validPlacement)
+                continue;
+
+            player.mines.push_back(Mine(x, y));
```
#could `player.mines.emplace_back(x, y);`

didierrenardgb commented on minefield_game/init.cpp:
```diff
@@ -35,69 +35,89 @@ namespace Init
         for (size_t i = 0; i < player.mines.size(); ++i)
         {
-            std::cout << "Mina #" << i + 1 << ": (" << player.mines[i].cell.getX() << "," << player.mines[i].cell.getY() << ")\n";
+            std::cout << "Mine #" << i + 1 << ": (" << player.mines[i].cell.getX() << "," << player.mines[i].cell.getY() << ")\n";
```
#should use parenthesis on the addition, `std::cout << "Mine #" << (i + 1) << ...`

didierrenardgb commented on minefield_game/init.cpp:
```diff
@@ -35,69 +35,89 @@ namespace Init
+        std::cout << "Each player has " << mines << " mines per turn\n";
     }
 
     void initializeGrid(Board &board)
```
#nice use of `resize` below!

didierrenardgb commented on minefield_game/player_header.h:
```diff
@@ -5,9 +5,11 @@ struct Mine;
+
 struct Mine
 {
     Cell cell;
```
#could rename this field such that this strong type makes more sense. i.e. `Cell location`, then you'll refer to it as `mine.location`, which looks nice!

didierrenardgb commented on minefield_game/statePattern/EndGameState.cpp:
```diff
@@ -0,0 +1,25 @@
+#include "GameContext.h"
+#include "EndGameState.h"
+#include "PlacementState.h"
+#include "../board_header.h"
```
#must not use relative paths this way. Instead, supply the proper include path to the build command. I.e. `clang -I. *.cpp -o minefield.exe`. The `-I` option is used to specify an include path (you can specify many by repeating the argument, i.e. `-Ifolder1 -Ifolder2`, etc.), and `.` points to current directory. Then you can remove the `../`, and it'd be better if you use `<>` in that case (i.e. `#include <board.h>`)

didierrenardgb commented on minefield_game/statePattern/GameContext.cpp:
```diff
@@ -0,0 +1,15 @@
+#include "GameContext.h"
+#include "GameState.h"
+
+void GameContext::setState(std::unique_ptr<GameState> state)
+{
+    currentState = std::move(state);
```
#nice usage of `std::unique_ptr` and move semantics!

didierrenardgb commented on minefield_game/statePattern/GameContext.h:
```diff
@@ -0,0 +1,23 @@
+class GameContext
+{
+private:
+    std::unique_ptr<GameState> currentState;
```
#nmi why making this `private` and everything else `public`? Moreso considering that there seem to be no invariance for this field to justify the encapsulation - aside from it, would look like the `GameContext` could have been a `struct` instead (pretty much as in [the cell case](https://github.com/ornella-mansilla/Minefield_Repository/pull/8/files#r2207916592)).

didierrenardgb commented on minefield_game/statePattern/GuessingState.cpp:
```diff
@@ -0,0 +1,85 @@
+#include "../init_header.h"
+#include <iostream>
+
+void GuessingState::handle(GameContext& context)
```
#should break this one down into smaller functions and/or methods, it's huge!

didierrenardgb commented on minefield_game/statePattern/GuessingState.cpp:
```diff
@@ -0,0 +1,85 @@
+    std::cout << "Player 1: You have " << guesses1 << " guesses.\n";
+    while (context.player1.guesses.size() < static_cast<size_t>(guesses1))
+    {
+        int x;
+        int y;
```
#should initialize variables to some value

didierrenardgb commented on minefield_game/statePattern/GuessingState.cpp:
```diff
@@ -0,0 +1,85 @@
+    {
+        int x;
+        int y;
+        std::cout << "Guess #" << context.player1.guesses.size() + 1 << " - Enter X: ";
```
#should parenthesize arithmetical expression, `(context.player1.guesses.size() + 1)`. That's because sometimes operator precedence could complicate things... not the particular case, but the compiler could have interpreted it this way:

```cpp
std::cout << "Guess #" << context.player1.guesses.size() + (1 << "- Enter X: ");
```

Since pointer types (in this case, the `char const* const` that is `"- Enter X: `) can be interpreted as numbers as they hold memory addresses inside, we could have ended with 1 being shifted left (multiplied by 2) a very high number of times, ending up in a message like "Guess #5646817987654" when in fact it was the first one. It is not the case, but it's better to build the habit of parenthesizing those kind of things such that you can be sure it won't happen.

didierrenardgb commented on minefield_game/statePattern/GuessingState.cpp:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    std::cout << "\nPlayer 2: You have " << guesses2 << " guesses.\n";
```
#should make a function/method for this as it's essentially the same implementation as above, just changing the affected player - that cries for an abstraction of some sort!

didierrenardgb commented on minefield_game/statePattern/ResolutionState.cpp:
```diff
@@ -0,0 +1,159 @@
+    // save destroyed mines
+    for (const auto &oldMine : defender.mines)
+    {
+        bool wasDestroyed = std::none_of(updatedMines.begin(), updatedMines.end(), [&](const Mine &m)
+                                         { return cellMatches(m.cell, oldMine.cell); });
```
#nice usage of `<algorithm>` and lambdas!

But... #should not capture everything (that's what `[&]` is doing), just capture what you need instead (`[&oldMine](const Mine& m) { return cellMatches(m.cell, oldMine.cell); }`)

didierrenardgb commented on minefield_game/statePattern/ResolutionState.cpp:
```diff
@@ -0,0 +1,159 @@
+
+    defender.mines = updatedMines;
+
+    // self-damage: check if the player guessed where he had a mine
```
#nice use of a comment that adds value

didierrenardgb replied:
But... probably #should move into its own function/method which could have an equally expressive name 😅

didierrenardgb commented on minefield_game/statePattern/ResolutionState.cpp:
```diff
@@ -0,0 +1,159 @@
+        }
+    }
+
+    for (const auto &m2 : p2.mines)
```
#should abstract it into a function and reuse for both players

ornella-mansilla committed:
const por parametros homogeneos

ornella-mansilla committed:
cambios en resolution state, player y cell.h

ornellalujan committed:
cambios en dev pull request de dev a devBranch

cambios en dev pull request

ornellalujan committed:
cambios en resolution state + ver setup state

ornellalujan committed:
cambios a resolution state + ver game setup state (merge from devBranch to dev)

cambios a resolution state + ver game setup state

ornella-mansilla committed:
refactoring + ver el print del tablero y el placement de las minas

ornella-mansilla committed:
cambios por el must + constexpr

ornella-mansilla committed:
cambios en convencion de parametros con const + print tablero

ornella-mansilla committed:
cambios finales

ornella-mansilla committed:
cambios finales 2

ornella-mansilla committed:
Merge second code review #10 from ornella-mansilla/dev

second code review

ornella-mansilla committed:
cambios: bounds checking + constexpr convention

luroblesgb reviewed (approved):
Very good!

didierrenardgb commented on minefield_game/board.cpp:
```diff
@@ -1,50 +1,66 @@
+        {
+            int mineX = mine.location.x;
+            int mineY = mine.location.y;
+            if (mineX == board.axisX[x] && mineY == board.axisY[y])
```
#must do bounds checking before accessing

didierrenardgb commented on minefield_game/board.cpp:
```diff
@@ -1,50 +1,66 @@
+
+    void printRow(Board const& board, Player const& player, size_t y)
+    {
+        int displayY = board.axisY[y];
```
#must do bounds checking before accessing

didierrenardgb commented on minefield_game/init.cpp:
```diff
@@ -45,72 +50,88 @@ namespace Init
+        int indexX = x - 1;
+        int indexY = y - 1;
+
+        if (board.grid[indexY][indexX].isCellTaken())
```
#must do bounds checking before accessing

ornella-mansilla committed:
cambios: board sin axis, solo grid + nombre del archivo main

ornella-mansilla committed:
Merge bounds checking + constexpr convention#11 from ornella-mansilla/dev

bounds checking + constexpr convention

ornella-mansilla committed:
cambios: board sin axis, solo grid + nombre del archivo main.cpp

ornella-mansilla committed:
Remove axes from Board; unify coordinate handling via grid only; main cpp file´s name changed #12 from ornella-mansilla/dev

Remove axes from Board; unify coordinate handling via grid only; main cpp file´s name changed

luroblesgb reviewed (approved):
OK!
